Wiki instruments (#7555)
aces/Loris: docs/wiki/01_STUDY_PARAMETERS_SETUP/02_Clinical_Instruments/01_instrument_intro_prerequisites.md

mode is enabled, a new table must be created and data is stored into the dedicated 
instrument table.

>_**Note:** The support for JSON data storage is a new implementation that is meant 
> to deprecate the use of SQL tables in the medium to long term. It is currently 
> offered as an option only but will eventually replace the SQL method._

## Prerequisites

1. Prepare your **Case Report Forms (CRFs)** including any “look-up” tables used for 
scoring/normalization and any documentation defining possible relationships between 
fields for validation.
2. Determine which **sections of the questionnaire should be included or omitted** 
from the online version. Typically, unless they pertain to the patients themselves, 
administrative details included on the original form can be ignored, but instructions 
that will be helpful for the examiner/patient might be included. 
3. Decide **how all your fields would map to the form element** types allowed by LORIS
(Refer to the Instrument Builder module to get a list of field types available for LORIS forms)
4. Determine if your instrument requires **special certifications** to be administered. 

> If so, only certified examiners will be able to sign the instrument. (*For more 

Reviewer: What does "sign the instrument" mean?

LORIS users from the same site and the same project. LORIS allows for instrument 
specific access permissions if needed. (*For more information, see [Additional Configurations - Permissions](04_instrument_additional_configurations.md#Permissions) section*)
6. Determine if your instrument will be **administered by an examiner** and Data-entry 
will be done by personnel **OR** if the instrument will be completed by the **candidate 
directly**.(*For more information, see [Additional Configurations - Surveys](04_instrument_additional_configurations.md#Surveys) section*)
7. Decide which **Quality Assurance (QA)** features you would like to use for the 
instrument [Double Data Entry](05_instrument_quality_assurance.md#Double-Data-Entry) 
and/or [Validity Flagging](05_instrument_quality_assurance.md#Validity)

## Instrument Development Workflow

_This section is meant to describe the recommended workflow of developing and 
installing an instrument in the database. The next sections in the documentation 
follow the same order as the steps listed below._

1. Read the entire "instrument" documentation before getting started.

> 2. Determine your answer for every in the [Prerequisites](#Prerequisites) section above

Reviewer: "for every in the"?

in a given study. These behavioural, psychological, and clinical tests are known as 
**instruments** or instrument forms.  

Basic instrument forms can be complemented with automated **scoring** and **validation**. 
The implementation of scoring and validation varies based on the chosen instrument 
format, the instructions below describe in more details how the implementation should 
be done for each type.

The **battery** is a list of which instruments are to be administered on each cohort 
of participants, and typically varies per timepoint and per study site.  For example, 
the battery of instruments administered at a participant's first visit (e.g. timepoint 
V01) may depend on which cohort and study site they belong to.  A participant may be 
tested on a certain set of instruments if they are in cohort A and from site X, but 
would be tested on different instruments if they are registered in cohort B at site Y. 
For certain studies, the battery may also depend on the age of the participant. 
(For more information about Battery assignments, navigate directly to the 
[Clinical Test Battery](./07_clinical_test_battery.md) section or continue following 

> this order of the).  

Reviewer: I think something to cut off here. "or continue following this order of the)."?

### How are Instruments Created?

LORIS instruments can be implemented in one of 2 formats: **LINST** or **PHP**.

**LINST** is a custom LORIS form format which is generated and edited by the LORIS 
Instrument Builder module. The LINST format allows for the creation of basic forms 
and can be complemented by a *score* file to allow for automated scoring  and a *rules* 
file to accommodate for validation of dependencies between form questions and answers.

**PHP** instruments are forms coded manually by a developer in the PHP language. 
These instruments can also have rules and scoring defined.

Choosing the format in which an instrument is coded is necessary before beginning 

> the process. PHP instruments can eventually be converted to LINST but the opposite is not 

Reviewer: I don't think this statement is true.. PHP instruments can do arbitrary code which might not be compatible with LINST?

Author: valid... and there is no reason to include that anyways

true. Once an instrument is implemented in LINST format, transforming it to PHP is not 
trivial. Here are a few bullet points that can help you make your decision.

- ***LINST***
- *Pros:*
    - Easy to create (LORIS Instrument Builder)
    - Ideal for simple instruments
    - Less prone to bugs and issues
    - Can be complimented with rules and scoring 
- *Cons:*
    - Validation logic can be limited in complex cases

>     - Validation and scoring might still require a developer

Reviewer: That's not a con of LINST, it's true of both formats.

Author: true but I do think it should be mentioned somewhere that ther eis no way of making rules and scoring without a developer

    - Limited flexibility of form elements (default LORIS form elements are used)
- ***PHP***
- *Pros:*
    - Very flexible as form elements can be overwritten, implemented or customized
    - Allows for more complex validation logic
    - Can be converted back to LINST (Manual intervention is only required for validation and scoring)
- *Cons:*
    - Requires a developer
    - Significant maintenance required over time to adapt to new LORIS releases
    - less stable as some elements might change get deprecated over time

> _**Note:** The definitions below can help choosing an instrument format._
> - _**Validation/Rules:** LORIS allows for 3 levels of form validation in instruments. 
>    Level 1 consists in setting a form field as required or not to the completion 
>    of the form. Level 2 set a field to required or not based on the response of 
>    another field in the form. Level 3 uses custom validation functions to determine 
>    the validity or the value set for a field. LINST instruments can use levels 1 

> >    and 2, only PHP instruments can use level 3 validation._

Reviewer: I don't understand where these "levels" came from or what this is trying to say.

Author: Right....

These "levels" are a way for me to describe the different kind of validation we can do in forms... I might have not explained it in depth enough, maybe this should live in it's own document not mixed with the "how to code an istrument" but basically:

Level 1: you can make a field required or not... not based on any other field or anything
Level 2: you make a field required based on the value of another field
Level 3 (only in PHP instruments): you can do any sort of validation (not limited to making a field required or not) by making a validate function in the instrument...

Reviewer 2: @ridz1208 maybe "format" instead of level?

Author: I don't mind the name as long as the difference between the "tiers"/"levels"/"options"/"formats" is clear

Reviewer: My concern is more that if this is just a concept that's introduced in this documentation someone might email the mailing list/open a bug report and developers won't know what they're talking about than the specific word used.

> > - _**Scoring:** Scoring in LORIS instruments is always done using the PHP coding 

Reviewer: Scoring can be done in any language for LINST instruments--it just executes a command in the instrument directory.

Author: good catch

7. Decide which **Quality Assurance (QA)** features you would like to use for the 
instrument [Double Data Entry](05_instrument_quality_assurance.md#Double-Data-Entry) 
and/or [Validity Flagging](05_instrument_quality_assurance.md#Validity)

## Instrument Development Workflow

_This section is meant to describe the recommended workflow of developing and 
installing an instrument in the database. The next sections in the documentation 
follow the same order as the steps listed below._

1. Read the entire "instrument" documentation before getting started.
2. Determine your answer for every in the [Prerequisites](#Prerequisites) section above
3. Start Developing your instrument either using the [**LINST**](./02_instrument_format/LINST_instrument.md) 
or [**PHP**](./02_instrument_format/PHP_instrument.md) format.
4. Install your instrument by following the [Installation Instructions](03_instrument_install.md)
5. Configure your new instruments following the [Additional Configurations](04_instrument_additional_configurations.md) section
6. Finally, [set up your behavioural test battery](07_clinical_test_battery.md) in 

> 7. order to be able to [test and troubleshoot](06_instrument_testing_and_troubleshooting.md) your instrument 

Reviewer: The "in order" being split between 6/7 flows really weirdly.

Author: I'll try to amend

Author: it was just a wrong order formatting

Basic instrument forms can be complemented with automated **scoring** and **validation**. The implementation of scoring and validation varies based on the chosen instrument format, the instructions below describe in more details how the implementation should be done for each type.

The **battery** is a list of which instruments are to be administered on each cohort of participants, and typically varies per timepoint and per study site.  For example, the battery of instruments administered at a participant's first visit (e.g. timepoint V01) may depend on which cohort and study site they belong to.  A participant may be tested on a certain set of instruments if they are in cohort A and from site X, but would be tested on different instruments if they are registered in cohort B at site Y (For more information about Battery assignments, navigate directly to the [Clinical Test Battery](./07_clinical_test_battery.md) section or continue following this order of the).

### How are Instruments Created?

LORIS instruments can be implemented in one of 2 formats: **LINST** or **PHP**.

> **LINST** is a LORIS in-house form format which is generated and edited by the LORIS Instrument Builder module. The LINST format allows for the creation of basic forms but can be complemented by a *score* file to allow for automated scoring and a *rules* file to accommodate for automatically checked rules.

Reviewer 3: ```suggestion
**LINST** is a custom LORIS form format which is generated and edited by the LORIS Instrument Builder module. The LINST format allows for the creation of basic forms with limited functionality and can be complemented by a *score* file to allow for automated scoring and a *rules* file to accommodate for validation of dependencies between form questions and answers.
```
- changing in-house to "custom" (i think in-house won't make as much sense to externals)
- changing "rules" wording to match the very nice intro on this file which emphasizes their purpose: validation

Author: will commit with a slight mod. I removed the `limited functionality` as it sounds pejorative here

> **PHP** instruments are forms coded manually by a developer in the PHP language. These instruments can also have rules and scoring defined.

Reviewer 3: ```suggestion
**PHP** instruments are forms coded manually by a developer in the PHP language. These instruments can also have rules and scoring defined as well as more complex form logic, within a single PHP file. Norms tables can also be referenced using PHP format instruments. 
```
I think it's good to mention norms tables here, and complex form logic.

Choosing the format in which and instrument is coded is necessary before beginning the process. PHP instruments can easily be converted to LINST but the opposite is not true. Once an instrument is implemented in LINST format, transforming it to PHP is not trivial. Here are a few bullet points that can help you make your decision.

- ***LINST***
 - *Pros:*
     - Easy to implement (LORIS Instrument Builder)
     - Ideal for small instruments
     - Less prone to bugs in the long term 
     - Can be combined with rules and scoring 
 - *Cons:*
     - Rules and scoring logic can be limited in complex cases
     - Rules and scoring might still require a developer
     - Does not offer flexibility (default LORIS form elements are used)
- ***PHP***
 - *Pros:*
     - Very flexible as form elements can be overwritten, implemented or customized
     - allows for more complex validation and scoring

>      - Can easily be converted back to LINST

Reviewer 3: ```suggestion
     - Can easily be converted back to LINST
     - Can be coded with complex branching logic and to reference norms tables
```

Author: @christine isnt that redundant with the second point `allows for more complex validation and scoring`

Author: okai I'll put it somewhere in there !

 - *Cons:*
     - Requires a developer
     - Significant maintenance required over time to adapt to new LORIS releases
     - less stable as some elements might change get deprecated over time

### How is Instrument Data Stored?

The data entered by users in an instrument form can be stored in 2 different formats: **SQL table** or **JSON string**. 

JSON data storage makes instrument installation easier by removing the need for table creation in the database and reducing maintenance efforts in the long term. However, SQL storage allows for data amendments to be done directly through SQL commands without the need to parse the JSON data using a tool.

The JSON data string is stored in the `Data` column of the `flag` table. If the SQL mode is enabled, a new table must be created and data is stored into the dedicated instrument table.

> >_**Note:** The support for JSON data storage is a new implementation that is meant to deprecate the use of SQL tables in the long term but is currently offered as an alternative only._

Reviewer 3: ```suggestion
>_**Note:** The support for JSON data storage is a new implementation that is meant to deprecate the use of SQL tables in the long term but is currently offered as an alternative only._
```
so... is it not fully supported?  can we just say it's in Beta testing? 
Are there any modules it doesn't automatically work with like Data Dictionary / DQT?

Author: as far as I know all modules support it right now. I would like to stay clear from offering as beta because projects will avoid using it if its labeled as such. It has been in beta since 23 and I think at this stage its actually an alternate

Reviewer 3: Ok makes sense.

> ## Prerequisites

Reviewer 3: If we don't provide templates, this is really just Best Practices or Suggested steps.  
Some sort of visual, like a decision tree or flow chart would help to clarify the path through both the php/linst decision and these steps, no?

Author: yes, how do we do that in markdown?

a screen capture ?

> 1. Prepare your **Case Report Forms (CRFs)** including any “lookup” tables used for scoring/normalization and any documentation defining possible relationships between fields for validation.

Reviewer 3: ```suggestion
1. Prepare your **Case Report Forms (CRFs)** including any “lookup” tables used for scoring/normalization and any documentation defining possible relationships between fields for validation.
```
Please add & LInk a template in the repo?   Case Report Form isn't standard for a lot of projects/users

Author: I might remove that point then ? we dont have a CRF example as far ass i know

Author: well... what do projects usually use for instruments ?? I thought everyone used CRFs as a base... its the questionnaire

2. Determine which **sections of the questionnaire should be included or omitted** from the online version. Typically, unless they pertain to the patients themselves, administrative details included on the original form can be ignored, but instructions that will be helpful for the examiner/patient might be included. 

> 3. Decide **how all your fields would map to the form element** types allowed by LORIS

Reviewer 3: example template would help here, if this is truly a pre-requisite. Especially for creating LINST forms this is going to be important (given the choices between types of date fields, planning rules, etc. )

Author: I will link to existing forms

4. Determine if your instrument requires **special certifications** to be administered. If so, only certified examiners will be able to sign the instrument. (*For more information, see [Additional Configurations - Certification](04_instrument_additional_configurations.md#Certification) section*)
5. Determine if your instrument will have **special access restrictions** other than LORIS users from the same site and the same project. LORIS allows for instrument specific access permissions if needed. (*For more information, see [Additional Configurations - Permissions](04_instrument_additional_configurations.md#Permissions) section*)

> 6. Determine if your instrument will be **administered by an examiner** and Data-entry will be done by personnel **OR** if the instrument will be completed by the **candidate directly**.(*For more information, see [Additional Configurations - Surveys](04_instrument_additional_configurations.md#Surveys) section*)

Reviewer 3: Some sort of a decision tree would be useful here

> 7. Decide which **Quality Assurance (QA)** features you would like to use for the instrument [Double Data Entry](05_instrument_quality_assurance.md#Double-Data-Entry) and/or [Validity Flagging](05_instrument_quality_assurance.md#Validity)

Reviewer 3: After this line, give people a brief summary of the rest of the instrument installation process: 
- Then, make the instrument and configure the battery 
- install and test the instrument
- deploy

Author: but that is later in the documentation not here

Author: I'm re-reading this and I think you mean outside of the prerequisite section? add a new section for installation summary section ? is that it?

Reviewer 3: Yes, that would be helpful - a roadmap to instrument installation since it's such a long/forking process
